### ISSUE TITLE
build_targets generate duplicate inconsistent targets

### ISSUE THREAD

**Issue by superaha** [ultralytics/yolov5#2593]
## ❔Question

I am trying to look understand how build_targets works. I deep dive into the code to help me. However, I find that the current build_targets can generate duplicate targets but matched to different bounding boxes. In other words, some locations in the feature map may be used to compute inconsistent bbox regression or bbox cls targets.

Here is an example:

```    
data = torch.stack((b.cuda(), a.cuda(), gj.cuda(), gi.cuda()), dim = 1)
data_u, idx, cnts = torch.unique(data, dim = 0, return_counts = True, return_inverse = True)
```

then
```
print(cnts)
tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1], device='cuda:0')
```
The concatenation of b, a, gi, and gj, in my understanding, should be unique, aka, at each location each anchor is responsible to learn one ground-truth bbox. Here, we see some will need to learn two targets.

Let's take one duplicate example as follows:

```
dups = data_u[cnts > 1,:]
rows = torch.where( (data == dups[idup,:].unsqueeze(0)).all(dim = 1))[0]
print(rows)
tensor([ 84, 158], device='cuda:0')
```
And, the data and the gt-box look like:
```
print(t[[84,158],:])
tensor([[ 1.00000,  2.00000,  7.07629, 22.01196,  1.60349,  0.83117],
        [ 1.00000,  2.00000,  5.66685, 22.22872,  1.94376,  0.77795]], device='cuda:0')
print(data[[84,158],:])
tensor([[ 1,  0, 22,  6],
        [ 1,  0, 22,  6]], device='cuda:0')
```

So basically, this example says: for the second image in this batch, the first anchor at the feature map location [22,6] is responsible to regress two bounding box [ 7.07629, 22.01196,  1.60349,  0.83117] and [5.66685, 22.22872,  1.94376,  0.77795].

This may cause instable gradient in the learning.

Let me know what you guys think whether my understanding is correct or not.

**Comment by glenn-jocher** [comment]
@superaha your analysis seems roughly correct, except I disagree that this might cause loss instability.

ML training naturally sums multiple losses to create a summed or average gradient for each parameter. We are always summing losses (with competing interests) across anchors, grid points, and types (box, obj, cls). This is simply adding additional losses using the same process.

In the end there is no right answer, simply different paths people take to try to arrive at the same destination. The above process favors recall, while isolating matches to single pairs would favor precision. Since ML is an empirical science, I would recommend you train with any proposed modifications you have against the default code to really understand their effects, and of course if you manage to produce improvements above the baseline on common datasets (VOC, COCO), please let us know!

**Comment by lyf6** [comment]
@superaha I got the same question with you. It's really confused.
